Disable eth_sign by default, allow users to toggle it back on

409H opened the pull request:
## Explanation

After looking at various phishing kits ([and de-obfuscating](https://twitter.com/sniko_/status/1587882460976713728)), there are enormous trends between different kits that they all abuse `eth_sign` to ask users to inadvertently sign a transaction (as MetaMask UI is opaque to an on-chain transaction being signed). You can read more about the attack [here](https://officercia.medium.com/how-hackers-may-steal-your-ethers-and-why-does-eth-sign-function-matter-69497ee11593).

I'd estimate that the abuse of `eth_sign` is near equal to the abuse of Wyvern and Seaport contracts in a manner to steal NFTs through creating orders for 0 eth to a specific beneficiary.

In short: `eth_sign` has been consistently abused by phishers/scammers to steal funds and the case may be that only power-users who use very legacy dapps may need to still use `eth_sign`.

* Disable `eth_sign` by default - throw an `ethErrors.rpc.methodNotFound` exception
* Allow the user to enable `eth_sign` through "Advanced Settings"

RPC Method preferences are stored within `rpcMethodPreferences` in the store state, allow for us to develop this further if someone wants to lock down rpc calls into MetaMask

## Screenshots/Screencaps
![Screenshot 2023-01-20 at 17 12 30](https://user-images.githubusercontent.com/2313704/213761789-4656cd56-e2e6-4274-ab31-0b0fc03f3d11.png)

## Manual Testing Steps

1. Go to https://metamask.github.io/test-dapp/
2. Open Developer Tools
3. Click "Sign" button under the heading "Eth Sign"
4. See the exception thrown and no prompt to sign within MetaMask
5. Go into MetaMask > Settings > Advanced Settings
6. Toggle eth_sign to be enabled
7. `goto: 1`

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

409H committed:
Added translation for eth sign toggle

409H committed:
Disabled the ability to call eth_sign by default. Added ability within advanced settings to renable support for eth_sign

409H committed:
Add test case for eth_sign being enabled and disabled

holantonela commented:
The content of the toggle should be:

`Title:` Enable `eth_sign`
`Description:` Turn this on to enable dapps to request you `eth_sign` requests. `eth_sign` is an open-ended signing method that allows signing an arbitrary hash, which means it can be used to sign transactions, or any other data, making it a dangerous phishing risk. Only sign `eth_sign` requests if you are able to read what you are signing and you trust the origin.
`Default:` OFF (?)

@coreyjanssen could we have your review here?

coreyjanssen commented:
revised content here (made it a bit more user-friendly):

`Title`: Allow `eth_sign` requests

`Description`: Turn this on to let dapps request your signature using `eth_sign` requests. `eth_sign` is an open-ended signing method that lets you sign an arbitrary hash, making it a dangerous phishing risk. Only sign `eth_sign` requests if you can read what you are signing and trust the origin of the request.

`Default`: OFF

@holantonela

409H committed:
Modified copy

409H commented:
@holantonela @coreyjanssen thank you, i have pushed the copy change in https://github.com/MetaMask/metamask-extension/pull/17308/commits/37d0ba3bd2482e3d054a408478bfa647a8e04060

409H committed:
Moved rpc method preference to its own object within store

409H committed:
Complete work on moving rpc method preference into its own object within store

409H committed:
Fix with prettier

409H committed:
Fix lint

409H committed:
Fix a unit test

409H committed:
Fix test

holantonela commented:
> I think if an eth_sign request comes to extension we should show message to user to enable it in settings, else it may be broken experience for those we are using it currently.

Good point. But, if we ask users to enable it the right away, we are missing the point of making it optional. After we release, we can learn more things about how users use or not `eth_sign`. 

In the case we want to prompt users to change defaults, we can offer users the option to save the origin domain in a safe list so if they trust the origin, they allow them to trigger `eth_sign` requests even when it is disabled (and other perks too). This should be a local list managed in Settings > General by all the accounts.

409H commented on app/scripts/controllers/preferences.js:
```diff
@@ -31,6 +31,9 @@ export default class PreferencesController {
       dismissSeedBackUpReminder: false,
+      rpcMethodPreferences: {
+        ethSign: false,
+      },
```
What would you suggest instead?

409H replied:
Renamed it in https://github.com/MetaMask/metamask-extension/pull/17308/commits/aece3d9398b1caadbead00addd0d2384350f7e0d

409H committed:
Renamed function and object keys to be more intuitive

409H committed:
Fix e2e test

409H committed:
Merge remote-tracking branch 'origin/develop' into feature/toggle-ethsign

* origin/develop: (142 commits)
  Use network provider state, instead of CurrencyRateController state, to select 'nativeCurrency' (#17450)
  [e2e]Add e2e test for deleting custom network (#17254)
  Fix #17362 - Ensure NFT icons are correct (#17440)
  Fix #17385 - Provide autohide callback for ActionMessage (#17392)
  button link housekeeping (#16885)
  Bump lavapack version to 5 (#17431)
  feature: convert shared/modules/contract-utils.test.js to typescript (#17435)
  Refactor confirm page container component from class to functional component (#16907)
  Test parameter change cache miss (#17346)
  Document callback-style background calls as deprecated (#17376)
  Improving getMemoizedMetadataContractName selector (#17432)
  [GridPlus] Bumps `gridplus-sdk` to v2.4.1 (#16847)
  Put hardware wallets behind an HARDWARE_WALLETS_MV3 flag (#17354)
  Consolidated all component import paths (#17368)
  updated new links for the stories (#17369)
  Fix #17388 - Remove dismiss button from NFT notification (#17389)
  Remove a notification for falling back from STX to regular swaps (#17374)
  Feat/15086/add banner severities component (#17307)
  fix icon names (#17391)
  Fix Playwright install step (#17415)
  ...

jpuri reviewed (dismissed):
Looks good 👍

seaona commented on test/e2e/tests/eth-sign.spec.js:
```diff
@@ -31,10 +59,18 @@ describe('Eth sign', function () {
         await driver.fill('#password', 'correct horse battery staple');
         await driver.press('#password', driver.Key.ENTER);
 
+        // Enable eth_sign
```
Instead of enabling `eth_sign` with Settings, we could skip all the steps below by enabling it from the state. This will make the test faster and complies with the standards followed on e2e.

Detailed explanation: on `fixture-builder` file, we could add the state for `eth_sign` and set it to `true` on a function below, in the same way it's done with other preferences.

Then on this test, we can just enable that fixture like this example:

```
.withPreferencesController({
            preferences: {
              showFiatInTestnets: true,
            },
```

409H replied:
Thanks @seaona - changed in https://github.com/MetaMask/metamask-extension/pull/17308/commits/c51da5e223c179e096ddd96fced8416bd827d486

409H committed:
Enabled eth_sign through preferences fixture

409H committed:
Merge branch 'develop' into feature/toggle-ethsign

409H committed:
Fix lint

Commit:
Fix e2e test

Wait for the notification popup to close, leaving 2 window handles the extension and the test dapp

Commit:
Fix e2e test

Commit:
Fix unit test

Enable eth_sign method

danjm committed:
Merge branch 'develop' into feature/toggle-ethsign

danjm committed:
Lint fix

409H committed:
Merge branch 'develop' into feature/toggle-ethsign
